quickwit-common/src/uri.rs

```diff
@@ -18,46 +18,97 @@
-        Some((protocol, path)) => (protocol, path.to_string()),
-    };
-
-    let current_dir = env::current_dir().context("Could not fetch the current directory.")?;
-
-    if protocol == FILE_PROTOCOL {
-        if path.starts_with('~') {
-            let home_dir_path = home::home_dir()
-                .context("Could not fetch the home directory.")?
+const PROTOCOL_SEPARATOR: &str = "://";
+
+/// Encapsulates the URI type.
+#[derive(Debug, PartialEq, Eq)]
+pub struct Uri {
+    uri: String,
+    protocol_idx: usize,
+}
+
+impl Uri {
+    /// Tries to to construct a Uri from the raw string.
+    /// A `file://` protocol is assumed if not specified.
+    /// File URIs are resolved (normalised) relative to the current working directory
+    /// unless an absolute path is specified.
+    /// Handles special characters like (~, ., ..)
+    pub fn try_new(uri: &str) -> anyhow::Result<Self> {
+        let (protocol, mut path) = match uri.split_once(PROTOCOL_SEPARATOR) {
+            None => (FILE_PROTOCOL, uri.to_string()),
+            Some((protocol, path)) => (protocol, path.to_string()),
+        };
+
+        if protocol == FILE_PROTOCOL {
+            let current_dir =
+                env::current_dir().context("Could not fetch the current directory.")?;
+
+            if path.starts_with('~') {
+                let home_dir_path = home::home_dir()
+                    .context("Could not fetch the home directory.")?
+                    .to_string_lossy()
+                    .to_string();
+
+                let user_name_dir = format!("~{}", whoami::username());
+                let user_name_dir_prefix = format!("~{}/", whoami::username());
+                if path == user_name_dir || path == "~" {
+                    path = home_dir_path;
+                } else if path.starts_with(&user_name_dir_prefix) {
+                    path.replace_range(0..user_name_dir.len(), &home_dir_path);
+                } else if path.starts_with("~/") {
+                    path.replace_range(0..1, &home_dir_path);
+                }
+            }
+
+            if !path.starts_with('/') {
+                path = current_dir.join(path).to_string_lossy().to_string();
+            }
+
+            path = normalize_path(Path::new(&path))
                 .to_string_lossy()
                 .to_string();
-
-            path.replace_range(0..1, &home_dir_path);
         }
 
-        if !path.starts_with('/') {
-            path = current_dir.join(path).to_string_lossy().to_string();
+        Ok(Self {
+            uri: format!("{}{}{}", protocol, PROTOCOL_SEPARATOR, path),
+            protocol_idx: protocol.len(),
+        })
+    }
+
+    /// Returns the uri protocol.
+    pub fn protocol(&self) -> &str {
+        &self.uri[..self.protocol_idx]
+    }
+
+    /// Returns the uri path.
+    /// If protocol is `file://`, returns [`Some(Path)`] otherwise returns [`None`].
+    pub fn path(&self) -> Option<&Path> {
```

Comment:
Maybe we should rename it filepath? (path is also the name of a component in an uri.)

Reply:
:+1: Yes and that's why I was suggesting it should always return something rather than Option. But with `filepath` it's very clear what we want.